sdcflows/transform.py

```diff
@@ -72,151 +274,271 @@ def fit(self, spatialimage):
-            newshape = spatialimage.shape
+        # TODO Separate cache validation from here. With the resampling, this
+        # code will always  determine the cache must be recalculated.
+        # if self.mapped is not None:
+        #     newshape = target_reference.shape
 
-            if np.all(newshape == self.mapped.shape) and np.allclose(
-                newaff, self.mapped.affine
-            ):
-                return False
+        #     if np.all(newshape == self.mapped.shape) and np.allclose(
+        #         target_affine, self.mapped.affine
+        #     ):
+        #         return False
 
         weights = []
         coeffs = []
+        _tmp_reference = None
+
+        hdr = target_reference.header.copy()
+        if approx:
+            from sdcflows.utils.tools import deoblique_and_zooms
+
+            # Generate a sampling reference on the fieldmap's space that fully covers
+            # the target_reference's grid.
+            target_reference = deoblique_and_zooms(
+                listify(self.coeffs)[-1],
+                target_reference,
+            )
 
         # Generate tensor-product B-Spline weights
         for level in listify(self.coeffs):
-            xfm.reference = spatialimage
-            moved_cs = level.__class__(
-                level.dataobj, xfm.matrix @ level.affine, level.header
-            )
-            wmat = grid_bspline_weights(spatialimage, moved_cs)
+            wmat = grid_bspline_weights(target_reference, level)
             weights.append(wmat)
             coeffs.append(level.get_fdata(dtype="float32").reshape(-1))
 
-        # Interpolate the VSM (voxel-shift map)
-        vsm = np.zeros(spatialimage.shape[:3], dtype="float32")
-        vsm = (np.squeeze(np.hstack(coeffs).T) @ sparse_vstack(weights)).reshape(
-            vsm.shape
+        # Reconstruct the fieldmap (in Hz) from coefficients
+        fmap = np.zeros(target_reference.shape[:3], dtype="float32")
+        fmap = (np.squeeze(np.hstack(coeffs).T) @ sparse_vstack(weights)).reshape(
+            fmap.shape
         )
 
-        # Cache
-        hdr = spatialimage.header.copy()
-        hdr.set_intent("estimate", name="Voxel shift")
+        # Generate a NIfTI object
+        hdr.set_intent("estimate", name="fieldmap Hz")
         hdr.set_data_dtype("float32")
-        hdr["cal_max"] = max((abs(vsm.min()), vsm.max()))
-        hdr["cal_min"] = - hdr["cal_max"]
-        self.mapped = nb.Nifti1Image(vsm, spatialimage.affine, hdr)
+        hdr["cal_max"] = max((abs(fmap.min()), fmap.max()))
+        hdr["cal_min"] = -hdr["cal_max"]
+
+        # Cache
+        self.mapped = nb.Nifti1Image(fmap, target_reference.affine, hdr)
+
+        if approx:
+            from nitransforms.linear import Affine
+
+            _tmp_reference = nb.Nifti1Image(
+                np.zeros(_tmp_shape, dtype=target_header.get_data_dtype()),
+                target_affine,
+                target_header,
```

Comment:
Does the dtype of the data array matter?

```suggestion
                np.zeros(target_reference.shape[:3], dtype=target_reference.get_data_dtype()),
                target_reference.affine,
                target_reference.header,
```